### PR TITLE
[Product Block Editor]: fallback with random products when no related products

### LINKED ISSUE #42922: Manage `Choose products for me`
**Prerequisites (mark completed items with an [x]):**
- [ ] I have checked that my issue type is not listed here https://github.com/woocommerce/woocommerce/issues/new/choose
- [ ] My issue is not a security issue, support request, bug report, enhancement or feature request (Please use the link above if it is).

**ACs**

```[tasklist]
### Tasks
- [x] When the user clicks Choose products for me, we add five products using the existing [Related Products](https://woo.com/document/related-products-up-sells-and-cross-sells/#section-3) functionality. AkVNGImLgSqCObTQ3idVn7-fi-12744_192579. Depends on #42849
- [x] If the user don’t have tags / attributes etc added yet, then we choose 5 randomly selected products that are publicly published and in-stock.
- [x] If a store has less than 5 other products that are publicly published and in-stock, then we choose the maximum number of products they have.
- [x] If there are products already selected under Upsell / Cross-sell (regardless whether they are user-selected ones or auto-generated ones), they all get replaced with the new recommended suggestions.
```

### PR DESCRIPTION
**Pull request by retrofox**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This allows to suggest random products in the Linked Products tab when the current product doesn't have related ones, as requested in the issue:

> If the user don’t have tags / attributes etc added yet, then we choose 5 randomly selected products that are publicly published and in-stock.

Also, it documents the getRelatedProducts() helper [function](https://github.com/woocommerce/woocommerce/blob/ea58165654d9a322f321654d9721c8846d361212/packages/js/product-editor/src/utils/get-related-products/Readme.md). 

Closes https://github.com/woocommerce/woocommerce/issues/42922

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Ensure your site has products that have defined categories
2. Enable the Linked Product flag
  - Enable the WCA plugin
  - Click on `Tools` -> `WCA Test Helper`
  - Click on `Features` tab
  - Toggle on the `product-linked` feature 

Screenshot 2023-12-26 at 11 36 05

3. Enable the new Product Block editor
4. Create a new Product
5. Go to the Linked Products tab
6. Ensure the product doesn't have defined related_ids:

```es6
wp.data.select( 'core' ).getEntityRecord( 'postType', 'product', [ <product-id> ] )?.related_ids
```

8. Click on the `Choose products for me` button of the Upsells or Cross-sells sections 
9. **Before (trunk)** Confirm the app doesn't provide any linked products 
10. **After (this PR)** Confirm the suggests five random products

https://github.com/woocommerce/woocommerce/assets/77539/42defd5a-ffc2-4b62-b30a-d05f59c7163c

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

File: packages/js/product-editor/src/utils/get-related-products/Readme.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+```es6
+getRelatedProducts( productId, options )
+```
+
+#### Parameters
+
+- `productId` (number): The ID of the product for which related products are to be fetched.
+- `options` (Object): An object containing the following property:
+  - `fallbackToRandomProducts` (boolean): Optional. If set to `true`, the function will return random products if no related products are found. Default is `false`.
```
Comment (octaedro) [inline packages/js/product-editor/src/utils/get-related-products/Readme.md]: I think that [the linter doesn't like](https://github.com/woocommerce/woocommerce/actions/runs/7506782506/job/20438956914?pr=43584#step:8:35) the tab here. Maybe we can change the `-` for a `+`

File: packages/js/product-editor/src/utils/get-related-products/Readme.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import getRelatedProducts from './path-to-getRelatedProducts';
+
+getRelatedProducts( 123, { fallbackToRandomProducts: true } )
+  .then( relatedProducts => {
+    console.log( relatedProducts );
+  } )
+  .catch( error => {
+    console.error( 'Error fetching related products:', error );
+  } );
+```
```
Comment (octaedro) [inline packages/js/product-editor/src/utils/get-related-products/Readme.md]: Could you [add an `Enter` here](https://github.com/woocommerce/woocommerce/actions/runs/7506782506/job/20438956914?pr=43584#step:8:36)?

File: packages/js/product-editor/src/utils/get-related-products/Readme.md
Comment (octaedro) [inline packages/js/product-editor/src/utils/get-related-products/Readme.md:1]: Thank you for adding the `README.md`, I have the feeling that a lot of times we forget to document this kind of thing and it's very valuable to do it 🙏

**Comment by octaedro** [comment]
Not directly related to this issue, I realized we're not blocking the ["Choose products for me" button](https://github.com/woocommerce/woocommerce/pull/43584/files#diff-9e1741bb3b5f53f4d4e44589c09c72d250b1cae57d52e69dd9e8997daa5d0d7aR166) when loading. Maybe we could add a blocker to prevent users from sending multiple requests when loading.

File: packages/js/product-editor/src/utils/get-related-products/index.ts [hunk 3]
```diff
@@ -4,20 +4,65 @@
...
+
+const POSTS_NUMBER_TO_RANDOMIZE = 30;
+
+/**
+ * Return related products for a given product ID.
+ * If fallbackToRandomProducts is true,
+ * return random products if no related products are found.
+ *
+ * @param {number}  productId - Product ID.
+ * @param {boolean} options   - Options.
```
Comment (octaedro) [inline packages/js/product-editor/src/utils/get-related-products/index.ts]: I think that `options` is an object, not a boolean.

File: packages/js/product-editor/src/utils/get-related-products/index.ts [hunk 4]
```diff
@@ -4,20 +4,65 @@
...
+		if ( ! lastPost?.length ) {
+			return;
+		}
+
+		const lastPostIds = lastPost.map( ( post ) => post.id );
+
+		// Pick five random post IDs
+		relatedProductIds = lastPostIds
+			?.sort( () => Math.random() - 0.5 )
+			.slice( 0, 5 );
```
Comment (octaedro) [inline packages/js/product-editor/src/utils/get-related-products/index.ts]: Just a small suggestion, what do you think about adding a `const NUMBER_OF_POSTS_TO_PICK = 5;`? That would make this value more readable and easy to maintain.

**Review by octaedro — COMMENTED** [review]
Good job @retrofox! This is testing well on my end and the code looks mainly good. I just left a couple of small suggestions.

File: packages/js/product-editor/src/utils/get-related-products/index.ts (hunk 3 above)
Comment (retrofox) [reply packages/js/product-editor/src/utils/get-related-products/index.ts]: Thanks. Actually we can use the `getRelatedProductsOptions` type

File: packages/js/product-editor/src/utils/get-related-products/index.ts (hunk 4 above)
Comment (retrofox) [reply packages/js/product-editor/src/utils/get-related-products/index.ts]: When I was working on it, I had the same idea. Let me update it.

File: packages/js/product-editor/src/utils/get-related-products/Readme.md (hunk 1 above)
Comment (retrofox) [reply packages/js/product-editor/src/utils/get-related-products/Readme.md]: 92b0f07cbfc42eb96e8578535603ef3a6a5a17d2

File: packages/js/product-editor/src/utils/get-related-products/Readme.md (hunk 2 above)
Comment (retrofox) [reply packages/js/product-editor/src/utils/get-related-products/Readme.md]: 👍 92b0f07cbfc42eb96e8578535603ef3a6a5a17d2

File: packages/js/product-editor/src/utils/get-related-products/index.ts (hunk 3 above)
Comment (retrofox) [reply packages/js/product-editor/src/utils/get-related-products/index.ts]: 92b0f07cbfc42eb96e8578535603ef3a6a5a17d2

File: packages/js/product-editor/src/utils/get-related-products/index.ts (hunk 4 above)
Comment (retrofox) [reply packages/js/product-editor/src/utils/get-related-products/index.ts]: 969c93d327183a78c568e99f7d4183aea742febc

**Comment by retrofox** [comment]
> Not directly related to this issue, I realized we're not blocking the https://github.com/woocommerce/woocommerce/pull/43584/files#diff-9e1741bb3b5f53f4d4e44589c09c72d250b1cae57d52e69dd9e8997daa5d0d7aR166 when loading. Maybe we could add a blocker to prevent users from sending multiple requests when loading.

Yes. Let's do it in a follow-up. We talked a little bit about it here p1705073474634869-slack-C06C7C3AQHL

**Comment by retrofox** [comment]
> Good job @retrofox! This is testing well on my end and the code looks mainly good. I just left a couple of small suggestions.

Thanks, Fer. I think all addressed

**Review by octaedro — APPROVED** [review]
Thank you @retrofox for addressing the changes. This is testing well here, but I think there already are [a couple of lint errors](https://github.com/woocommerce/woocommerce/actions/runs/7529178665/job/20492874721?pr=43584). I'm going to pre-approve this PR to merge it after fixing the problems.

**Comment by retrofox** [comment]
> Thank you @retrofox for addressing the changes. This is testing well here, but I think there already are [a couple of lint errors](https://github.com/woocommerce/woocommerce/actions/runs/7529178665/job/20492874721?pr=43584). I'm going to pre-approve this PR to merge it after fixing the problems.

They aren't blockers, Fernando. So let's merge it.

**Pull request merged**